README
Rewrite from pyeco.

Guide:
* add "local,127.0.0.1:13000," to server.lst
* run virtualeco_launcher
* register user on http://127.0.0.1:13100/
* copy and run pyecolauncher

Support:
saga17 (until ver 361)
361+ is unsupported because mapserver login packet changed.
If you can dump packet please help me,
I can't do that because I'm not in jp.

No more feature will be add because developer can't
get any more resource of packet and script.

-------------------------------

LICENSE: The MIT License (MIT)
Additions:
Use of this software is limited to private use of the individual or circle, use for commercial purposes is not allowed.
このソフトウェアは個人またはサークルの私的使用目的での利用に限ります、商用目的での利用は認めません。

-------------------------------

If you can dump packet from official server, please contact me.
Or if you dont known how to dump, I can provide tool to you.

Contact:
pyeco.git@gmail.com

Mail-List(Google Group):
https://groups.google.com/group/pyeco



ChangeLog
2013-04-26
	add ver 361+ unsupported message

2013-03-28
	update database to 361

2013-02-21
	update database to 359
	fix bug: pet not visible in some time (359+?)

2013-01-24
	update database to 356
	packet fixes with 356

2012-12-28
	packet fixes (still continue attack after monster dead, wait fix)

2012-12-27
	update database to 354
	353+ packet fixes
	fix wrong password when delete character cause client frozen
	support effect type equip
	update item category script

2012-12-21
	update database to 353
	many packet changed, wait fix.
	add additions to license.

2012-11-26
	update pet pict script
	client 351+ texture missing, still wait fix.

2012-11-22
	update database to 351

2012-11-05
	update item category script
	add shutdown command

2012-10-18
	update database to 348

2012-10-13
	add script.skill_add, skill_del, skill_clear
	add command /skill_add skill_id, /skill_del skill_id, skill_clear
	add script.printallskill, command /printallskill
	modify DEFAULT_GMLEVEL to 255
	require input to exit when except main script exception in launcher
	replace monsters *magic* to *skill* (physical skill packet same as magic)
	add test skill

2012-10-11
	add pack/io_unpack array, pack/unpack/io_unpack long
	move general.pack*/unpack*/io_unpack*, packet.pack* to packet_struct
	add test skill
	fix /effect error

2012-10-10
	fix warp coord error when 11f8(move) came after 05e6(event)
	pop pc on usermap when master logout

2012-10-09
	fix freeze in use skill call from item
	add lib.usermaps, lib.obj.usermap, pc.usermap_obj
	add script.lock_move, script.unlock_move
	limited support set/unset flygarden rope (only rope)
	DATABASE_FORMAT_VERSION 1.1.2: set item.skill_id, item.skill_id_add

2012-10-05
	add kuji category script
	add script.npcshop_list(pc, shop_item_list)
	add common pet pict switch script
	skip load file under script/ filename startswith __
	fix packet 09e9 send wrong pict id

2012-10-04
	add nekomata heart switch script
	add pc.item_event_id
	DATABASE_FORMAT_VERSION 1.1.1:
	petid->pet_id, eventid->event_id

2012-09-27
	add item, warehouse stock limit
	log file use utf-8 not system encoding
	move general.secure* to lib.security
	fix ctrl+c exit error on windows

2012-09-26:
	add script/site_packages
	import warp point script from yutori server
	move lib.env to lib._env (_env store default values)
	allow user replace their own lib/env.py each update
	move settings from server.ini to lib.env
	remove lib/obj/serverconfig.py, lib.server.config
	add script/loginevent.py (send after login if set env.SEND_LOGIN_EVENT)
	add npctrade id
	fix skill use error when x or y > 127

2012-09-25
	remove unnecessary lock from lib.server

2012-09-24
	add pet thread

2012-09-23
	let pet motion refer master (sit and stand)
	replace thread.start_new_thread to threading.Thread

2012-09-21
	support set and save pet_pict_id and pet_weapon_id
	fix script.wing set wrong attribute problem
	modify some enc parameter

2012-09-20
	update db to 346
	make pet heart visible
	save item data with dumpobj
	add lib.env.USERDATA_FORMAT_VERSION
	add lib.obj.pc_data_lib to load mutil-version pc data
	add lib.users.upgrade_user_data

2012-09-17
	limited support dem parts and form change
	add command /printallequip and script.printallequip
	add command /unsetallequip and script.unsetallequip
	remove pc.dic (str:str dictionary)
	add pc.var (str:object dictionary, support save and restore standard python object)
	add lib.skills
	add test skill heal, arrow

2012-09-14
	move lib.version to lib.env
	move a part of settings to lib.env
	move license from BSD to MIT

2012-09-01
	add lib.version
	pass rijndael_obj instead of rijndael_key to general.encode/decode
	add MAX_USER_ID, MAX_PC_ID, MAX_PET_ID, MAX_MONSTER_ID
	fix monster.spawn/delete error
	use general.make_id to generate user and pc id when create
	remove serveraddress from server.ini
	add serverbindip to server.ini
	add serverpublicip to server.ini
	check address before bind on windows

2012-08-26
	use system encoding for all log
	move general.Null to general.NullClass
	remove lib/unit_test (no longer use)

2012-08-24
	add type info in database row map (not use _detect_type any more)
	add version info in database dump
	database format version 1.1.0

2012-08-23
	add script.petmotion and command /petmotion motion_id motion_loop
	add script.petmotion_loop and command /petmotion_loop motion_id

2012-08-23
	update db to 345
	add dbmap, change database dump format from raw_list to raw_dict {id: __dict__}
	(auto generate attribute location and boost startup speed)
	add script.size and command /size pc_size
	add script.petstandby_on and command /petstandby_on
	add script.petstandby_off and command /petstandby_off

2012-08-02
	make dump files load and save with pyc format (if DUMP_WITH_ZLIB off)

2012-07-26
	update db to 343
	fix pc look sync problem

2012-07-21
	support drop and pick up item
	add general.make_id
	move monster.send_map to script.send_map
	add script.send_map_obj(map_obj, without_list, *args)
	add script.send_server(*args)
	add PC.item_append, item_pop, warehouse_append, warehouse_pop with system message
	remove script.item_object_add (PC.item_append instead of)
	clean up code

2012-07-20
	support pc to pc trade (unstable)
	fix script.setgold, npcmotion, npcshop, npcsell affect other pc on the same map

2012-07-17
	add general.secure_open replace __builtin__.open
	add general.secure_listdir replace os.listdir
	add general.secure_remove replace os.remove
	add general.secure_rmdir replace os.rmdir
	add general.secure_mkdir replace os.mkdir
	disable os.link, os.symlink
	add general.secure_rename replace os.rename
	add general.secure_chdir
	add general.secure_chmod replace os.chmod
	add general.secure_chown replace os.chown
	disable os.chdir, os.chroot, os.mkfifo


lib/_env.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import sys
import os
import time
#static
NAME = "virtualeco"
LAST_UPDATE = "2013-04-26"
DATABASE_FORMAT_VERSION = "1.1.3"
USERDATA_FORMAT_VERSION = "1.1.0"

#runtime
STARTUP_TIME = time.time()
LOAD_STARTUP_TIME = 0
SYSTEM = os.name
RUNTIME_VERSION = sys.version_info
RUNTIME_VERSION_ALL = sys.version
STDOUT = sys.stdout
STDERR = sys.stderr
SYSTEM_ENCODING = sys.getfilesystemencoding()

#config
USE_LOGFILE = False
USE_DEBUGER = True
BACKUP_USER_DATA_EVERY_DAY = False
MAX_CONNECTION_FROM_ONE_IP = 3
DUMP_WITH_ZLIB = False
ZIP_COMPRESS = False

#server config
LOGIN_SERVER_PORT = 13000
MAP_SERVER_PORT = 13001
WEB_SERVER_PORT = 13100
SERVER_BIND_ADDR = "0.0.0.0"
SERVER_BROADCAST_ADDR = "localhost"
DEFAULT_GMLEVEL = 255
LOGIN_EVENT_ID = 30
SEND_LOGIN_EVENT = True
MAX_ITEM_STOCK = 100
MAX_WAREHOURSE_STOCK = 500
SHUTDOWN_CONFIRM_WORD = "-y"
GMLEVEL_MAP = {
	"run": 100,
	"help": 0,
	"reloadscript": 255,
	"reloadsinglescript": 255,
	"shutdown_server": 255,
	"user": 0,
	"say": 0,
	"msg": 0,
	"warning": 0,
	"servermsg": 150,
	"where": 0,
	"warp": 100,
	"warpraw": 100,
	"update": 0,
	"hair": 50,
	"haircolor": 50,
	"face": 50,
	"wig": 50,
	"ex": 50,
	"wing": 50,
	"wingcolor": 50,
	"motion": 0,
	"motion_loop": 0,
	"item": 150,
	"printitem": 0,
	"takeitem": 150,
	"dustbox": 0,
	"warehouse": 100,
	"playbgm": 0,
	"playse": 0,
	"playjin": 0,
	"effect": 0,
	"speed": 100,
	"setgold": 150,
	"takegold": 150,
	"gold": 150,
	"npcmotion": 100,
	"npcmotion_loop": 100,
	"npcshop": 100,
	"npcsell": 100,
	"spawn": 150,
	"killall": 150,
	"emotion": 0,
	"emotion_ex": 0,
	"shownpc": 0,
	"hidenpc": 0,
	"blackout": 0,
	"whiteout": 0,
	"size": 0,
	"petstandby_on": 0,
	"petstandby_off": 0,
	"petmotion": 0,
	"petmotion_loop": 0,
	"unsetallequip": 0,
	"printallequip": 0,
	"printallskill": 0,
	"skill_add": 50,
	"skill_del": 50,
	"skill_clear": 50,
}

#path config
DEFAULT_BASE = "."
LOG_DIR = "./log"
STDOUT_LOG = "./log/%s.log"
STDERR_LOG = "./log/%s_error.log"
SERVER_CONFIG_PATH = "./server.ini"
DATABASE_DIR = "./data"
SCRIPT_DIR = "./script"
USER_DIR = "./user"
USER_BAK_DIR = "./user_bak"
USER_CONFIG_NAME = "user.ini"
PC_CONIG_NAME = "%d.ini"
WEB_DIR = "./web"
REG_USER_PAGE_PATH = "/index.html"
DEL_USER_PAGE_PATH = "/delete.html"
MODIFY_PASSWORD_PAGE_PATH = "/modify_password.html"